packages/realm/src/collection-accessors/Dictionary.ts

```diff
@@ -0,0 +1,143 @@
+      switch (value) {
+        case binding.ListSentinel: {
+          const accessor = createListAccessor<T>({ realm, itemType: binding.PropertyType.Mixed, typeHelpers });
+          return new indirect.List<T>(realm, dictionary.getList(key), accessor, typeHelpers) as T;
+        }
+        case binding.DictionarySentinel: {
+          const accessor = createDictionaryAccessor<T>({ realm, itemType: binding.PropertyType.Mixed, typeHelpers });
+          return new indirect.Dictionary<T>(realm, dictionary.getDictionary(key), accessor, typeHelpers) as T;
+        }
+        default:
+          return fromBinding(value) as T;
+      }
+    },
+    set(dictionary, key, value) {
+      assert.inTransaction(realm);
+
+      if (isJsOrRealmList(value)) {
+        dictionary.insertCollection(key, binding.CollectionType.List);
+        insertIntoListOfMixed(value, dictionary.getList(key), toBinding);
+      } else if (isJsOrRealmDictionary(value)) {
+        dictionary.insertCollection(key, binding.CollectionType.Dictionary);
+        insertIntoDictionaryOfMixed(value, dictionary.getDictionary(key), toBinding);
+      } else {
+        dictionary.insertAny(key, toBinding(value));
+      }
+    },
+  };
+}
+
+function createDictionaryAccessorForKnownType<T>({
+  realm,
+  typeHelpers,
+  isEmbedded,
+}: Omit<DictionaryAccessorFactoryOptions<T>, "itemType">): DictionaryAccessor<T> {
+  const { fromBinding, toBinding } = typeHelpers;
+  return {
+    get(dictionary, key) {
+      return fromBinding(dictionary.tryGetAny(key));
+    },
+    set(dictionary, key, value) {
+      assert.inTransaction(realm);
+
+      if (isEmbedded) {
+        toBinding(value, { createObj: () => [dictionary.insertEmbedded(key), true] });
+      } else {
+        dictionary.insertAny(key, toBinding(value));
+      }
+    },
+  };
+}
+
+/** @internal */
+export function insertIntoDictionaryOfMixed(
+  dictionary: Dictionary | Record<string, unknown>,
+  internal: binding.Dictionary,
+  toBinding: TypeHelpers["toBinding"],
+) {
+  // TODO: Solve the "removeAll()" case for self-assignment (https://github.com/realm/realm-core/issues/7422).
+  internal.removeAll();
+
+  for (const key in dictionary) {
+    const value = dictionary[key];
+    if (isJsOrRealmList(value)) {
+      internal.insertCollection(key, binding.CollectionType.List);
+      insertIntoListOfMixed(value, internal.getList(key), toBinding);
+    } else if (isJsOrRealmDictionary(value)) {
+      internal.insertCollection(key, binding.CollectionType.Dictionary);
+      insertIntoDictionaryOfMixed(value, internal.getDictionary(key), toBinding);
+    } else {
+      internal.insertAny(key, toBinding(value));
+    }
+  }
+}
+
+/** @internal */
+export function isJsOrRealmDictionary(value: unknown): value is Dictionary | Record<string, unknown> {
+  return isPOJO(value) || value instanceof indirect.Dictionary;
+}
+
+/** @internal */
+export function isPOJO(value: unknown): value is Record<string, unknown> {
```

Comment:
Not sure I understand what POJO is. Plain old JavaScript object? I guess this is Java world stuff?

Reply:
I'd prefer `isJsPlainObject` or at least a comment here

Reply:
I agree, but I'd rather do that in a different PR as this is merely copying code around.